samza-api/src/main/java/org/apache/samza/operators/descriptors/base/stream/StreamDescriptor.java

```diff
@@ -0,0 +1,142 @@
+            "streamId: %s, systemName: %s", streamId, systemName));
+    Preconditions.checkArgument(isValidId(systemName),
+        String.format("systemName must be non-empty and must not contain spaces or special characters. " +
+            "systemName: %s, streamId: %s", systemName, streamId));
+    Preconditions.checkArgument(serde != null,
+        String.format("Serde must not be null. streamId: %s systemName: %s", streamId, systemName));
+    Preconditions.checkArgument(systemDescriptor != null,
+        String.format("SystemDescriptor must not be null. streamId: %s systemName: %s", streamId, systemName));
+    Preconditions.checkArgument(systemDescriptor.getSystemName().equals(systemName),
+        String.format("System name in constructor: %s does not match system name in SystemDescriptor: %s. streamId: %s",
+            systemName, systemDescriptor.getSystemName(), streamId));
+    this.streamId = streamId;
+    this.systemName = systemName;
+    this.serde = serde;
+    this.systemDescriptor = systemDescriptor;
+  }
+
+  /**
+   * The physical name of the stream on the system on which this stream will be accessed.
+   * This is opposed to the {@code streamId} which is the logical name that Samza uses to identify the stream.
+   * <p>
+   * A physical name could be a Kafka topic name, an HDFS file URN, or any other system-specific identifier.
+   * <p>
+   * If not provided, the logical {@code streamId} is used as the physical name.
+   *
+   * @param physicalName physical name for this stream.
+   * @return this stream descriptor.
+   */
+  public SubClass withPhysicalName(String physicalName) {
+    this.physicalNameOptional = Optional.ofNullable(physicalName);
+    return (SubClass) this;
+  }
+
+  /**
+   * Additional system-specific properties for this stream.
+   * <p>
+   * These properties are added under the {@code streams.stream-id.*} scope.
+   *
+   * @param streamConfigs system-specific properties for this stream
+   * @return this stream descriptor
+   */
+  public SubClass withStreamConfigs(Map<String, String> streamConfigs) {
+    this.streamConfigs.putAll(streamConfigs);
+    return (SubClass) this;
+  }
+
+  public String getStreamId() {
+    return this.streamId;
+  }
+
+  public String getSystemName() {
+    return this.systemName;
+  }
+
+  public Serde getSerde() {
+    return this.serde;
+  }
+
+  public SystemDescriptor getSystemDescriptor() {
+    return this.systemDescriptor;
+  }
+
+  public Optional<String> getPhysicalName() {
+    return physicalNameOptional;
+  }
+
+  private boolean isValidId(String id) {
+    return StringUtils.isNotBlank(id) && ID_PATTERN.matcher(id).matches();
+  }
+
+  public Map<String, String> toConfig() {
+    HashMap<String, String> configs = new HashMap<>();
+    configs.put(String.format(SYSTEM_CONFIG_KEY, streamId), getSystemName());
+    getPhysicalName().ifPresent(physicalName ->
```

Comment:
No, any config resolution (fallbacks, defaults, etc.) will be handled as it is today - usually in the container when the configs are being read. This is necessary for backwards compatibility. Also, changing how configs are handled is out of scope for this SEP. We have a separate task to support using "streamId" instead of "system.stream" everywhere.